### DISCUSSION TITLE
Enable display of currencies other than USD for unit of account (Real, Rand, Quetzal, Sol)

### DISCUSSION THREAD

**Discussion by agbegin** [GaloyMoney/galoy#1280] [Ideas]
The Bitcoin Beach Wallet is being used in Guatemala and Peru to onboard communities to Bitcoin. Other projects have indicated that they would also use the wallet in their communities if they could see the value of sats in their local currency. 

What would it take for us to roll out a feature to enable users to pick their unit of account in settings? 

Currencies that we could look into incorporating first, based on existing usage or interest: 
- Guatemalan Quetzal (GTQ) 
- Peruvian Sol (PEN) 
- South African Rand (ZAR)
- Brazilian Real (BRL)

See examples from Muun.
![Muun-currency-selector](https://user-images.githubusercontent.com/90297599/166484941-404a3526-3921-4233-a8ba-93cf865615a5.PNG)
![Muun-currency-display-Quetzal](https://user-images.githubusercontent.com/90297599/166484944-a6b0a760-2f77-4f30-b7ad-ed1f3fc97b98.PNG)

What would it take to incorporate a currency selector in [galoy-mobile](https://github.com/GaloyMoney/galoy-mobile) and/or [web-wallet](https://github.com/GaloyMoney/web-wallet)? Is this something that could be issued as a bounty?

**Comment by sebastianwindeck** [comment]
Since the European system has an increasing awareness in the system it might be useful to add EUR 

What do you think?

**Reply by agbegin** [reply]
Definitely. Plan is to add every fiat currency (see comment below linking to the issue)

**Comment by agbegin** [comment]
Issue from @nicolasburtey: [display currency in every fiat currency #1202 ](https://github.com/GaloyMoney/galoy/issues/1202 )

Next step is to capture the implementation/thinking done so far and scope out the remaining work. 

cc @dolcalmi @vindard

**Comment by agbegin** [comment]
Sharing the currency conversion & POS work from [Fä Bu](https://twitter.com/fabu69420/status/1533625074904973312?s=20&t=Zf6otpLiiOozTlaLEd3uKQ) on twitter. 

[pagar.bitcoinlake.io](https://pagar.bitcoinlake.io/)

* Simple POS screen for paying to Bitcoin Beach Wallet username over lightning
* Ability to choose which currency to use as unit of account
* Can link directly to each user's receive page: https://pagar.bitcoinlake.io/agbegin

On github: https://github.com/fabu21/bbw-fiat-invoices

**Comment by vindard** [comment]
## Notes on implementing multiple display currencies

The idea for displaying multiple currencies in our system is as follows. There are 2 ways to support multiple currencies:
1. **_Naive_**
  Get the current rate for a given default currency and convert all balances, past transactions and receive/send amounts at the current price

1. **Persist transaction rates**
  Register each transaction against a given currency and rate at the time of transaction and store this currency/rate. Present transactions in the transaction history with their recorded currency and rate at the time of transaction. Get the current rate for a current default currency and convert all balances and receive/send amounts at the current price.

We plan to support the 2nd way, which would mean storing a single unique display currency and rate _at the time of transaction_. When a user parses their transaction history, they will see the transaction represented as the display currency and amount in that display currency at the time of transaction. This is for e.g. how Muun handles multiple currencies today.


## Support from backend

### Already implemented
- We have three fields that are currently populated for each transaction under `payment` transaction type:
    - `displayAmount`
    - `displayFee`
    - `displayCurrency`

  These fields currently are populated with only the USD price since we only support USD on the backend for now and do not have a multicurrency Price Service implemented as yet.

  These fields are only implemented for transactions with `payment` type so far (`displayFee` gets auto added as `0` if not present on all transactions).

  The following aggregation should confirm which transaction types otherwise get the correct value for these properties.
  ```javascript
  [
    {
      '$match': {
        'satsAmount': {
          '$exists': true
        }
      }
    }, {
      '$group': {
        '_id': '$type'
      }
    }
  ]
    ```



### TODO

**Data structures:**
- All transaction types should write to the 3 display currency fields in our db
  - [ ] figure out which transaction types are not populating the 3 new display currency fields as yet

  - [ ] refactor code to ensure the missing transaction types all write to these fields when new transactions are created

  - [ ] migrate old transactions to use these new fields

- Reading from these new fields when returning transaction info to the API

  The `Transaction` payload for the API still reads from `usd` and `feeUsd` fields for its display currency logic. We would need to:
  - [ ] refactor the `LedgerTransaction` logic to read from these new fields instead

  - [ ] include any missing selections in the `Transaction` graphql object to support multiple currencies

- Reading from these new fields for constructing messages for the Notifications Service
  - [ ] refactor notifications service `publish` events to use these new fields instead

**Multi-currency Price Service**

- Implement a new Price Service, or multi-currency support into our existing Price Service (@dolcalmi should know the status/plans for this work)

**Multicurrency handling in API**

- Support setting/getting a user-level "default display currency"

- Use a user's "default display currency" to fetch display currency data and record into the 3 display currency fields for each new transaction


## Frontend

For each transaction:
- detect a transaction's display currency
- parse a currency prefix (e.g. `$`, `AUD`)
- present the values for this currency from the `display...` fields on a transaction together with that currency's prefix

For balances:
- present balances in the current user's display currency
